sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +501,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
         // not(not(e))  =>  e
         case Not(e) => e
+        // De Morgan's law: !(a || b) => !a && !b
```

Comment:
@chenghao-intel The `Not` case added here is not for simplification. It's used to push `Not` predicates to the bottom, so that `e` in `Not(e)` cannot be either `And` or `Or`. CNF conversion code relies on this assumption.